### PR TITLE
DM-35098: MPU command multiplex, adds port statistics and utility functions

### PR DESCRIPTION
**Pull request by pkubanek**
Adds MPU multiplex, with 4 supporting commands (see README.md for details). Reports port statistics. Replaces waits with port read timeouts. Add commands to report port settings & statistics, reset errors, and reset counters.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,31 +21,62 @@ Add the library to your project. Create MPU object, and execute its Run method.
 You need to provide Command, Telemetry and Debug (Output) FIFOs, and port for
 commands. See [Example.vi](Example.vi) for details.
 
+## Command queue
+
+Command queue uses own multiplex to clear MPU error lines, program MPU or
+report port status. The command queue is checked for new command after run of
+a programmed loop (which is empty on statup). If command is available, it is
```
Comment (teweitsai) [inline README.md:28]: statup --> startup

File: README.md [hunk 2]
```diff
@@ -21,31 +21,62 @@ Add the library to your project. Create MPU object, and execute its Run method.
...
+executed, if not, programmed code is executed. This quarantees programming will
+not mess with program execution, as both cannot happen at the same time.
+
+| Code | Description                                            |
+| ---- | -------------------------------------------------------|
+| 0    | Clear error lines. If the program fails, error lines will be set. See unit telemetry for how to get current error lines status. |
+| 1    | Program MPU. Followed by single byte - program length, and the program itself. See below for supported instructions.            |
+| 2    | Report port telemetry to Telemetry FIFO.                                 |
+| 3    | Cleare unit telemetry counters.                                          |
```
Comment (teweitsai) [inline README.md:37]: Clear?

File: README.md [hunk 3]
```diff
@@ -21,31 +21,62 @@ Add the library to your project. Create MPU object, and execute its Run method.
...
 Telemetry command (30) dumps U8 values. Offset is calucalted from 0. This can
 be used to periodically dump measured values.
 
-## Port statistics
-
-Port statistics command (50) writes to debug output following values. Values
-are written as low endian 64 bits numbers (into U8 FIFO).
+## Unit telemetry
+
+After command 2 on command FIFO, unit telemetry is reported. The telemtetry
```
Comment (teweitsai) [inline README.md:63]: telemtetry -> telemetry

**Review by teweitsai — APPROVED** [review]
Looks good. I only checked the documentation part.

File: README.md (hunk 2 above)
Comment (pkubanek) [reply README.md:37]: Reset is better

**Review by pcortes80 — APPROVED** [review]
It looks good, based on my understanding of this project. It is not clear to me the criteria you use to manage errors on FIFOs, because in some VIs (e.g. _ExecuteInstruction.vi) you added error wires out of FIFO, and in others (e.g. _DumpTelemetry.vi) did not. I would suggest adding an explanation on VI Descriptions, controls, and indicators, to help other developers.

**Comment by pkubanek** [comment]
> It looks good, based on my understanding of this project. It is not clear to me the criteria you use to manage errors on FIFOs, because in some VIs (e.g. _ExecuteInstruction.vi) you added error wires out of FIFO, and in others (e.g. _DumpTelemetry.vi) did not. I would suggest adding an explanation on VI Descriptions, controls, and indicators, to help other developers.

The logic behind that is described in README.md - please see [5355349]. The short story is you don't want e.g. dumping of the telemetry fail because you failed before, as handling errors might kill you and lead in telemetry not reporting anything, which you would like to avoid.

**Pull request merged**